pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTestNew.java

```diff
@@ -0,0 +1,1543 @@
+      String segmentName = idealStateBuilder.getSegment(p, 0);
+      advanceASeqForPartition(idealState, segmentManager, partitionAssignment, segmentName, p, 1, 100, tableName);
+    }
+    idealState = idealStateBuilder.build();
+    validateLLCPartitionsIncrease(segmentManager, idealState, tableConfig, nPartitions, nReplicas, instances,
+        exceptionExpected, skipNewPartitions);
+
+    // increase num partitions
+    nPartitions = 10;
+    validateLLCPartitionsIncrease(segmentManager, idealState, tableConfig, nPartitions, nReplicas, instances,
+        exceptionExpected, skipNewPartitions);
+
+    // keep num partitions same - noop
+    validateLLCPartitionsIncrease(segmentManager, idealState, tableConfig, nPartitions, nReplicas, instances,
+        exceptionExpected, skipNewPartitions);
+
+    // keep num partitions same, but bad instances - error
+    instances = getInstanceList(1);
+    validateLLCPartitionsIncrease(segmentManager, idealState, tableConfig, nPartitions, nReplicas, instances,
+        exceptionExpected, skipNewPartitions);
+
+    // increase num partitions, but bad instances - error
+    nPartitions = 12;
+    skipNewPartitions = true;
+    validateLLCPartitionsIncrease(segmentManager, idealState, tableConfig, nPartitions, nReplicas, instances,
+        exceptionExpected, skipNewPartitions);
+
+    // increase num partitions, but disabled ideal state - noop
+    idealState = idealStateBuilder.disableIdealState().build();
+    exceptionExpected = false;
+    instances = getInstanceList(6);
+    validateLLCPartitionsIncrease(segmentManager, idealState, tableConfig, nPartitions, nReplicas, instances,
+        exceptionExpected, skipNewPartitions);
+  }
+
+  private void advanceASeqForPartition(IdealState idealState, FakePinotLLCRealtimeSegmentManager segmentManager,
+      PartitionAssignment partitionAssignment, String segmentName, int partition, int nextSeqNum, long nextOffset,
+      String tableName) {
+    String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+    LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+    segmentManager.updateOldSegmentMetadataZNRecord(tableName, llcSegmentName, nextOffset);
+    LLCSegmentName newLlcSegmentName =
+        new LLCSegmentName(rawTableName, partition, nextSeqNum, System.currentTimeMillis());
+    segmentManager.createNewSegmentMetadataZNRecord(tableName, newLlcSegmentName, nextOffset, partitionAssignment);
+    segmentManager.updateIdealStateOnSegmentCompletion(idealState, segmentName, newLlcSegmentName.getSegmentName(),
+        partitionAssignment);
+  }
+
+  private void validateLLCPartitionsIncrease(FakePinotLLCRealtimeSegmentManager segmentManager, IdealState idealState,
+      TableConfig tableConfig, int nPartitions, int nReplicas, List<String> instances, boolean exceptionExpected,
+      boolean skipNewPartitions) {
+    try {
+      Map<String, Map<String, String>> oldMapFields = new HashMap<>(idealState.getRecord().getMapFields());
```

Comment:
is this a deep copy? you can create a copy of the idealstate. i think there is a method to do that.